File: python/replace_elements_with_greatest_element_on_right_side.py
class Solution:
    # Time complexity O(n)
    def replaceElements(self, arr: list[int]) -> list[int]:
        rightMax: int = -1
        for i in reversed(range(len(arr))):
            newMax: int = max(rightMax, arr[i])
            arr[i] = rightMax
            rightMax = newMax

        return arr


def test_replaceElements_case_1():
    # arrange
    arr: list[int] = [17, 18, 5, 4, 6, 1]
    expected: list[int] = [18, 6, 6, 6, 1, -1]

    # act
    solution = Solution()
    actual = solution.replaceElements(arr)

    # assert
    assert expected == actual


def test_replaceElements_case_2():
    # arrange
    arr: list[int] = [400]
    expected: list[int] = [-1]

    # act
    solution = Solution()
    actual = solution.replaceElements(arr)

    # assert
    assert expected == actual

File: rust/src/code/replace_elements_with_greatest_element_on_right_side.rs
struct Solution {}

impl Solution {
    pub fn replace_elements(arr: Vec<i32>) -> Vec<i32> {
        let mut result: Vec<i32> = vec![0; arr.len() + 1];
        result[arr.len()] = -1;
        for i in (0..arr.len()).rev() {
            result[i] = std::cmp::max(result[i + 1], arr[i]);
        }

        result[1..].to_vec()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn replace_elements_case_1() {
        // arrange
        let arr: Vec<i32> = vec![17, 18, 5, 4, 6, 1];
        let expected: Vec<i32> = vec![18, 6, 6, 6, 1, -1];

        // act
        let actual = Solution::replace_elements(arr);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn replace_elements_case_2() {
        // arrange
        let arr: Vec<i32> = vec![400];
        let expected: Vec<i32> = vec![-1];

        // act
        let actual = Solution::replace_elements(arr);

        // assert
        assert_eq!(expected, actual);
    }
}
